SpokeStack/SpokeStack/AudioController.swift

```diff
@@ -0,0 +1,223 @@
+                             ioActionFlags,
+                             inTimeStamp,
+                             inBusNumber,
+                             inNumberFrames,
+                             UnsafeMutablePointer<AudioBufferList>(&bufferList))
+    if (status != noErr) {
+        return status
+    }
+
+    let data: Data = Data(bytes: buffers[0].mData!, count: Int(buffers[0].mDataByteSize))
+    
+    DispatchQueue.main.async {
+        AudioController.shared.delegate?.processSampleData(data)
+    }
+    
+    return noErr
+}
+
+class AudioController {
+    
+    // MARK: Public (properties)
+    
+    static let shared: AudioController = AudioController()
+    
+    weak var delegate: AudioControllerDelegate?
+    
+    var sampleRate: Int = 16000
+    
+    var bufferDuration: TimeInterval = 10
+    
+    // MARK: Private (properties)
+    
+    fileprivate var remoteIOUnit: AudioComponentInstance?
+    
+    lazy private var audioComponentDescription: AudioComponentDescription = {
+        
+        var componentDescription: AudioComponentDescription = AudioComponentDescription()
+        
+        componentDescription.componentType = kAudioUnitType_Output
+        componentDescription.componentSubType = kAudioUnitSubType_RemoteIO
+        componentDescription.componentManufacturer = kAudioUnitManufacturer_Apple
+        componentDescription.componentFlags = 0
+        componentDescription.componentFlagsMask = 0
+        
+        return componentDescription
+    }()
+    
+    // MARK: Initializers
+    
+    deinit {
+        AudioComponentInstanceDispose(remoteIOUnit!)
+    }
+    
+    // MARK: Public (methods)
+    
+    func startStreaming() -> Void {
+
+        /// Prepare
+        
+        do {
+        
+            try self.prepare()
+            self.start()
+
+        } catch AudioError.audioSessionSetup(let message) {
+        
+            self.delegate?.setupFailed(message)
+            
+        } catch AudioError.general(let message) {
+        
+            self.delegate?.setupFailed(message)
+            
+        } catch {
+        
+            self.delegate?.setupFailed("An unknown error occured setting the stream")
+        }
+    }
+    
+    func stopStreaming() -> Void {
+        self.stop()
+    }
+    
+    // MARK: Private (methods)
+    
+    @discardableResult
+    private func start() -> OSStatus {
+        return AudioOutputUnitStart(remoteIOUnit!)
+    }
+    
+    @discardableResult
+    private func stop() -> OSStatus {
+        return AudioOutputUnitStop(remoteIOUnit!)
+    }
+    
+    @discardableResult
+    private func prepare() throws -> OSStatus {
+        
+        var status: OSStatus = noErr
+        let session: AVAudioSession = AVAudioSession.sharedInstance()
+        
+        /// AVSession setup
+        
+        do {
+        
+            try AVAudioSession.sharedInstance().setCategory(.playAndRecord, mode: .spokenAudio, options: .defaultToSpeaker)
```

Comment:
Nice, I didn't know that `AVAudioSession` had a mode for speech. Of course there's the mysterious `voicePrompt` mode too…